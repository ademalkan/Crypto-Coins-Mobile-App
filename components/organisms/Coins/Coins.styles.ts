import { StyleSheet } from "react-native";

export default StyleSheet.create({
  container: {
    flex: 1
  },
  title: {
    margin: 12,
    fontWeight: "600",
    fontSize: 18,
    borderBottomWidth: 1,
    borderBottomColor: '#f2f2f2',
    paddingBottom: 4
  },
})
