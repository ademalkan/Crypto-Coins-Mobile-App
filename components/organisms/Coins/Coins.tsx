import React from 'react'
import AlertText from '../../atoms/AlertText/AlertText';
import { VirtualizedList, Text, ScrollView } from 'react-native';
import Coin from '../../molecules/Coin/Coin';
import styles from './Coins.styles'
import { CoinI } from '../../../interfaces/CoinInterface';

interface CoinsProps {
  searchQuery: string;
  emptyText: string;
  coins: any;
  title: string;
}

const Coins: React.FC<CoinsProps> = ({ coins, searchQuery, title = "", emptyText = "You have not find any coins yet" }) => {
  if (!coins || !coins.length) {
    return (
      <AlertText text={emptyText} />
    )
  }


  const CoinMemoized = React.memo(Coin);

  const renderItem = ({ item }: { item: CoinI }) => (
    <CoinMemoized
      name={item.name}
      symbol={item.symbol.toUpperCase()}
      price={item.current_price ? item.current_price : item.price}
      total_volume={item.total_volume}
      price_change_24h={item.price_change_24h}
      image={item.image}
      id={item.id}
    />
  );

  const filteredCoins = coins.filter(
    (coin: CoinI | undefined) =>
      coin?.name.toLowerCase().includes(searchQuery.toLowerCase()) ||
      coin?.symbol.toLowerCase().includes(searchQuery.toLowerCase())
  );

  const ITEM_HEIGHT = 100;

  return (
    <>
      {title && <Text style={styles.title}>{title}</Text>}
      {filteredCoins &&
        <VirtualizedList
          style={styles.container}
          data={filteredCoins}
          renderItem={renderItem}
          keyExtractor={(item: CoinI) => item.id}
          initialNumToRender={10}
          maxToRenderPerBatch={10}
          updateCellsBatchingPeriod={10}
          windowSize={10}
          removeClippedSubviews={true}
          getItem={(data, index) => data[index]}
          getItemCount={(data) => data.length}
          getItemLayout={(data, index) => ({
            length: ITEM_HEIGHT,
            offset: ITEM_HEIGHT * index,
            index,
          })}
        />
      }

    </>
  )
}

export default Coins