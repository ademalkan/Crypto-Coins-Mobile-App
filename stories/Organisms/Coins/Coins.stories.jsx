// import React from "react";
// import { View } from "react-native";
// import { storiesOf } from "@storybook/react-native";
// import Coins from "../../../components/organisms/Coins/Coins";

// const coins = [
//   {
//     id: "1",
//     name: "Bitcoin",
//     symbol: "BTC",
//     current_price: 50000,
//     total_volume: 1000000,
//     price_change_24h: 5,
//     image: "bitcoin.png",
//   },
//   {
//     id: "2",
//     name: "Ethereum",
//     symbol: "ETH",
//     current_price: 3000,
//     total_volume: 500000,
//     price_change_24h: -2,
//     image: "ethereum.png",
//   },
//   // Add more coin data here
// ];

// const searchQuery = "";
// const emptyText = "You have not find any coins yet";
// const title = "Top Coins";

// storiesOf("Organisms/Coins", module)
//   .addDecorator((Story) => (
//     <View style={{ flex: 1, justifyContent: "center", alignItems: "center" }}>
//       <Story />
//     </View>
//   ))
//   .add("Default", () => (
//     <Coins
//       coins={coins}
//       searchQuery={searchQuery}
//       title={title}
//       emptyText={emptyText}
//     />
//   ));
