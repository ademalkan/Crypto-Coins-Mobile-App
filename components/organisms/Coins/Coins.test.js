import React from "react";
import renderer from "react-test-renderer";
import Coins from "./Coins";

jest.mock("expo-router", () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

describe("Coins", () => {
  it("should render the coins component correctly", () => {
    const coins = [
      {
        id: "1",
        name: "Bitcoin",
        symbol: "BTC",
        current_price: 50000,
        total_volume: 1000000,
        price_change_24h: 5,
        image: "coin.png",
      },
      {
        id: "2",
        name: "Ethereum",
        symbol: "ETH",
        current_price: 3000,
        total_volume: 500000,
        price_change_24h: -2,
        image: "coin.png",
      },
    ];

    const searchQuery = "";
    const title = "Coin List";
    const emptyText = "You have not found any coins yet";

    const component = renderer.create(
      <Coins
        coins={coins}
        searchQuery={searchQuery}
        title={title}
        emptyText={emptyText}
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
