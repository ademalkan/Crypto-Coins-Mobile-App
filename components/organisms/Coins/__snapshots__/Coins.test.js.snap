// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coins should render the coins component correctly 1`] = `
[
  <Text
    style={
      {
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 1,
        "fontSize": 18,
        "fontWeight": "600",
        "margin": 12,
        "paddingBottom": 4,
      }
    }
  >
    Coin List
  </Text>,
  <RCTScrollView
    data={
      [
        {
          "current_price": 50000,
          "id": "1",
          "image": "coin.png",
          "name": "Bitcoin",
          "price_change_24h": 5,
          "symbol": "BTC",
          "total_volume": 1000000,
        },
        {
          "current_price": 3000,
          "id": "2",
          "image": "coin.png",
          "name": "Ethereum",
          "price_change_24h": -2,
          "symbol": "ETH",
          "total_volume": 500000,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={10}
    windowSize={10}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          style={
            {
              "borderBottomColor": "#f2f2f2",
              "borderBottomWidth": 1,
              "flex": 1,
              "flexDirection": "row",
              "marginBottom": 10,
              "padding": 10,
              "width": "100%",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <ViewManagerAdapter_ExpoImage
              height={70}
              marginRight={10}
              onError={[Function]}
              onLoad={[Function]}
              onLoadStart={[Function]}
              onProgress={[Function]}
              proxiedProperties={
                {
                  "contentFit": "cover",
                  "contentPosition": {
                    "left": "50%",
                    "top": "50%",
                  },
                  "placeholder": [],
                  "source": [],
                  "tintColor": undefined,
                  "transition": {
                    "duration": 1000,
                  },
                }
              }
              style={
                {
                  "height": 70,
                  "marginRight": 10,
                  "width": 70,
                }
              }
              width={70}
            />
          </View>
          <View>
            <Text
              style={
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginRight": 5,
                }
              }
            >
              Bitcoin
            </Text>
            <Text
              style={
                {
                  "color": "#666",
                  "fontSize": 14,
                  "textTransform": "uppercase",
                }
              }
            >
              BTC
            </Text>
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#666",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "marginBottom": 4,
                    "textTransform": "capitalize",
                  }
                }
              >
                Total Volume:
              </Text>
              <Text
                style={
                  {
                    "color": "#666",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "marginLeft": 4,
                  }
                }
              >
                $
                1000000
              </Text>
            </View>
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#666",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "marginBottom": 4,
                    "textTransform": "capitalize",
                  }
                }
              >
                Price Change 24 Hours:
              </Text>
              <Text
                style={
                  {
                    "color": "#666",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "marginLeft": 4,
                  }
                }
              >
                $
                5
              </Text>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "display": "flex",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginTop": 10,
                  "width": 205,
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "backgroundColor": "#383a68",
                    "borderRadius": 8,
                    "opacity": 1,
                    "paddingHorizontal": 16,
                    "paddingVertical": 8,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "white",
                      },
                      undefined,
                    ]
                  }
                >
                  Detail
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "backgroundColor": "#383a68",
                    "borderRadius": 8,
                    "opacity": 1,
                    "paddingHorizontal": 16,
                    "paddingVertical": 8,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "white",
                      },
                      undefined,
                    ]
                  }
                >
                  Add Favorite
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              {
                "fontSize": 16,
                "fontWeight": "bold",
                "marginLeft": "auto",
              }
            }
          >
            $
            50000
          </Text>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          style={
            {
              "borderBottomColor": "#f2f2f2",
              "borderBottomWidth": 1,
              "flex": 1,
              "flexDirection": "row",
              "marginBottom": 10,
              "padding": 10,
              "width": "100%",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <ViewManagerAdapter_ExpoImage
              height={70}
              marginRight={10}
              onError={[Function]}
              onLoad={[Function]}
              onLoadStart={[Function]}
              onProgress={[Function]}
              proxiedProperties={
                {
                  "contentFit": "cover",
                  "contentPosition": {
                    "left": "50%",
                    "top": "50%",
                  },
                  "placeholder": [],
                  "source": [],
                  "tintColor": undefined,
                  "transition": {
                    "duration": 1000,
                  },
                }
              }
              style={
                {
                  "height": 70,
                  "marginRight": 10,
                  "width": 70,
                }
              }
              width={70}
            />
          </View>
          <View>
            <Text
              style={
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginRight": 5,
                }
              }
            >
              Ethereum
            </Text>
            <Text
              style={
                {
                  "color": "#666",
                  "fontSize": 14,
                  "textTransform": "uppercase",
                }
              }
            >
              ETH
            </Text>
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#666",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "marginBottom": 4,
                    "textTransform": "capitalize",
                  }
                }
              >
                Total Volume:
              </Text>
              <Text
                style={
                  {
                    "color": "#666",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "marginLeft": 4,
                  }
                }
              >
                $
                500000
              </Text>
            </View>
            <View
              style={
                {
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#666",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "marginBottom": 4,
                    "textTransform": "capitalize",
                  }
                }
              >
                Price Change 24 Hours:
              </Text>
              <Text
                style={
                  {
                    "color": "#666",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "marginLeft": 4,
                  }
                }
              >
                $
                -2
              </Text>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "display": "flex",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginTop": 10,
                  "width": 205,
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "backgroundColor": "#383a68",
                    "borderRadius": 8,
                    "opacity": 1,
                    "paddingHorizontal": 16,
                    "paddingVertical": 8,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "white",
                      },
                      undefined,
                    ]
                  }
                >
                  Detail
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "backgroundColor": "#383a68",
                    "borderRadius": 8,
                    "opacity": 1,
                    "paddingHorizontal": 16,
                    "paddingVertical": 8,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "white",
                      },
                      undefined,
                    ]
                  }
                >
                  Add Favorite
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              {
                "fontSize": 16,
                "fontWeight": "bold",
                "marginLeft": "auto",
              }
            }
          >
            $
            3000
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
